benchmarks/halo_finder/ArborX_HaloFinder.hpp

```diff
@@ -0,0 +1,264 @@
+/****************************************************************************
+ * Copyright (c) 2012-2020 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HALO_FINDER_HPP
+#define ARBORX_HALO_FINDER_HPP
+
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Macros.hpp>
+
+#include <chrono>
+
+namespace ArborX
+{
+
+template <typename View>
+struct Wrapped
+{
+  View _M_view;
+  double _r;
+};
+
+template <typename View>
+auto wrap(View v, double r)
+{
+  return Wrapped<View>{v, r};
+}
+
+namespace Traits
+{
+template <typename View>
+struct Access<Wrapped<View>, PredicatesTag>
+{
+  using memory_space = typename View::memory_space;
+  static size_t size(Wrapped<View> const &w) { return w._M_view.extent(0); }
+  static KOKKOS_FUNCTION auto get(Wrapped<View> const &w, size_t i)
+  {
+    return attach(intersects(Sphere{w._M_view(i), w._r}), (int)i);
+  }
+};
+} // namespace Traits
+
+namespace HaloFinder
+{
+template <typename MemorySpace>
+struct CCSCallback
+{
+  using tag = ArborX::Details::InlineCallbackTag;
+  Kokkos::View<int *, MemorySpace> stat_;
+
+  KOKKOS_INLINE_FUNCTION
+  int representative(int const i) const
+  {
+    // ##### ECL license (see LICENSE.ECL) #####
+    int curr = stat_(i);
+    if (curr != i)
+    {
+      int next, prev = i;
+      while (curr > (next = stat_(curr)))
+      {
+        stat_(prev) = next;
+        prev = curr;
+        curr = next;
+      }
+    }
+    return curr;
+  }
```

Comment:
Per [1]
```
Note that the [`representative()`] code is re-entrant and synchronization free                                                                                                                                                                                                                                                                                                                                                                                                                                                                             even though concurrent execution may cause data races on the parent array.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
However, these races are guaranteed to be benign for the following reasons.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
First, the only write to shared data is in [`parent[prev] = next; `]. This                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
write updates a single aligned machine word and is therefore atomic. Moreover,                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
it overwrites a valid entry with another valid entry. Hence, it does not matter                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
if other threads see the old or the new value as either value will allow them                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
to eventually reach the representative. Similarly, all the reads of the parent                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
array will either fetch the old or new value, but both values are acceptable.                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
The only problem that can occur is that two threads try to update the same                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
parent pointer at the same time. In this case, one of the updates is lost. This                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
reduces the code’s performance as duplicate work is performed and the path is                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
not shortened by as much as it could have been, but it does not result in                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
incorrect paths. On average, the savings of not having to perform                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
synchronization far outweighs this small cost. Lastly, it should be noted that                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
the rest of the code either accesses the parent array via calls to the                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
find_repres function or changes the parent pointer of a representative vertex                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
but never of a vertex that is in the middle of a path. If the find_repres code                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
already sees the new representative, it will return it. Otherwise, it will                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
return the old representative. Either return value is handled correctly.
```
In short, it uses the fact that writing data to a single aligned machine word is atomic and that it does not matter whether the value is the latest one or not.